| Name                                      | ID           |
|-------------------------------------------|--------------|
| Ahmed Mostafa AbdEl-Rahman Hassan         | 20221372883  |


In [1]:
import pandas as pd
from sklearn.model_selection import train_test_split
from sklearn.naive_bayes import GaussianNB
from sklearn.metrics import confusion_matrix

C:\Users\Lenovo\Downloads\anaconda\lib\site-packages\scipy\__init__.py:146: UserWarning: A NumPy version >=1.16.5 and <1.23.0 is required for this version of SciPy (detected version 1.26.2
  warnings.warn(f"A NumPy version >={np_minversion} and <{np_maxversion}"


# Download Census Data

In [2]:
url = "file:///C:/Users/Lenovo/AppData/Local/Temp/f32a47bf-5db7-46ec-b40b-85943033637a_adult(1).zip.37a/adult.data"
columns = [
    "age", "workclass", "fnlwgt", "education", "education-num", "marital-status",
    "occupation", "relationship", "race", "sex", "capital-gain", "capital-loss",
    "hours-per-week", "native-country", "income"
]
data = pd.read_csv(url, names=columns, na_values=' ?', skipinitialspace=True)
data.head(5)

age         workclass  fnlwgt  education  education-num  \
0   39         State-gov   77516  Bachelors             13   
1   50  Self-emp-not-inc   83311  Bachelors             13   
2   38           Private  215646    HS-grad              9   
3   53           Private  234721       11th              7   
4   28           Private  338409  Bachelors             13   

       marital-status         occupation   relationship   race     sex  \
0       Never-married       Adm-clerical  Not-in-family  White    Male   
1  Married-civ-spouse    Exec-managerial        Husband  White    Male   
2            Divorced  Handlers-cleaners  Not-in-family  White    Male   
3  Married-civ-spouse  Handlers-cleaners        Husband  Black    Male   
4  Married-civ-spouse     Prof-specialty           Wife  Black  Female   

   capital-gain  capital-loss  hours-per-week native-country income  
0          2174             0              40  United-States  <=50K  
1             0             0              13  United-States  <=50K  
2             0             0              40  United-States  <=50K  
3             0             0              40  United-States  <=50K  
4             0             0              40           Cuba  <=50K

# Preprocess the Data

In [3]:
# Drop rows with missing values
data = data.dropna()
# Encode categorical variables
data = pd.get_dummies(data, columns=["workclass", "education", "marital-status", "occupation", "relationship", "race", "sex", "native-country"])
# Map income to binary values (0 for <=50K, 1 for >50K)
data['income'] = data['income'].map({'<=50K': 0, '>50K': 1})

# Split data into features (X) and target variable (y)
X = data.drop('income', axis=1)
y = data['income']

# Train Naive Bayes Classifier

In [4]:
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=42)

nb_classifier = GaussianNB()
nb_classifier.fit(X_train, y_train)


GaussianNB()

# Predict and Compute Sensitivity and Specificity

In [5]:
y_pred = nb_classifier.predict(X_test)
tn, fp, fn, tp = confusion_matrix(y_test, y_pred).ravel()
sensitivity = tp / (tp + fn)
specificity = tn / (tn + fp)
print("Sensitivity:", sensitivity)
print("Specificity:", specificity)

Sensitivity: 0.32017823042647997
Specificity: 0.9514366653176851


Sensitivity (True Positive Rate): This value (0.3202) suggests that the classifier correctly identifies about 32.02% of individuals who actually make over 50K a year from the total number of individuals who actually make over 50K a year.

Specificity (True Negative Rate): This value (0.9514) indicates that the classifier correctly identifies about 95.14% of individuals who actually do not make over 50K a year from the total number of individuals who do not make over 50K a year.

# Compute Posterior Probability

In [6]:
posterior_probs = nb_classifier.predict_proba(X_test)
positive_class_probs = posterior_probs[:, 1]  # Probability of income >50K
print("Posterior Probability of making over 50K a year:", positive_class_probs)
average_prob = positive_class_probs.mean()
print("Average posterior probability of making over 50K a year:", average_prob)

Posterior Probability of making over 50K a year: [4.31088775e-03 1.37859620e-02 1.71229441e-02 ... 1.00000000e+00
 6.39925511e-03 6.53278628e-04]
Average posterior probability of making over 50K a year: 0.12424877416671441


These values represent the probability assigned by the classifier to each instance belonging to the positive class (income >50K). The probabilities are scaled between 0 and 1, and they represent the confidence of the classifier in its predictions.